### PR TITLE
Changed entire system to use heading pitch instead of YPR

### PR DESCRIPTION
**Pull request by OsiJr**
- Removed speed limit on Zoom To
- Terrain is now picked using the standard picking pipeline (works with DEM and non-flat terrain)
- Rotate to uses quaternions
- Added more GIS helpers

### CODE REVIEW

File: src/imgui_ex/ImGuizmo.cpp [hunk 1]
```diff
@@ -284,7 +284,7 @@ static void vcGizmo_ComputeContext(const vcCamera *pCamera, const udDouble4x4 &m
   sGizmoContext.mModelInverse = udInverse(sGizmoContext.mModel);
   sGizmoContext.mMVP = sGizmoContext.camera.matrices.viewProjection * sGizmoContext.mModel;
 
-  sGizmoContext.mCameraDir = udDirectionFromYPR(sGizmoContext.camera.eulerRotation);
+  sGizmoContext.mCameraDir = udDirectionFromYPR(udDouble3::create(sGizmoContext.camera.headingPitch, 0.0)); //TODO: Fix me in ECEF
```
Comment (OsiJr) [inline src/imgui_ex/ImGuizmo.cpp:287]: This is logged as AB#1389

File: src/scene/vcPOI.cpp [hunk 2]
```diff
@@ -280,11 +280,11 @@ void vcPOI::AddToScene(vcState *pProgramState, vcRenderData *pRenderData)
     // Update the camera if the camera is coming along
     if (pProgramState->cameraInput.pAttachedToSceneItem == this && m_cameraFollowingAttachment)
```
Comment (OsiJr) [inline src/scene/vcPOI.cpp:281]: POIs haven't been fixed

Comment (OsiJr) [reply src/scene/vcPOI.cpp:281]: They have now

File: src/vcCamera.h [hunk 3]
```diff
@@ -21,7 +22,8 @@ enum vcCameraScrollWheelMode
 struct vcCamera
 {
   udDouble3 position;
-  udDouble3 eulerRotation;
+  //udDouble3 eulerRotation;
```
Comment (peterAdams1014) [inline src/vcCamera.h:25]: Just remove this now.

File: src/scene/vcViewpoint.cpp [hunk 4]
```diff
@@ -18,9 +18,16 @@ vcViewpoint::vcViewpoint(vdkProject *pProject, vdkProjectNode *pNode, vcState *p
 
 void vcViewpoint::OnNodeUpdate(vcState *pProgramState)
 {
-  vdkProjectNode_GetMetadataDouble(m_pNode, "transform.rotation.x", &m_CameraRotation.x, 0.0);
-  vdkProjectNode_GetMetadataDouble(m_pNode, "transform.rotation.y", &m_CameraRotation.y, 0.0);
-  vdkProjectNode_GetMetadataDouble(m_pNode, "transform.rotation.z", &m_CameraRotation.z, 0.0);
+  vdkError result;
+
+  result = vdkProjectNode_GetMetadataDouble(m_pNode, "transform.heading", &m_CameraHeadingPitch.x, 0.0);
+  result = vdkProjectNode_GetMetadataDouble(m_pNode, "transform.pitch", &m_CameraHeadingPitch.y, 0.0);
```
[the remark this answers is not in the document]
Comment (OsiJr) [reply src/scene/vcViewpoint.cpp:24]: Yeah

File: src/vcGIS.cpp [hunk 5]
```diff
@@ -94,3 +94,67 @@ udDouble3 vcGIS_GetWorldLocalUp(const vcGISSpace &space, udDouble3 localCoords)
...
+{
+  if (!space.isProjected || space.zone.projection >= udGZPT_TransverseMercator)
+    return udDouble3::create(0, 1, 0);
+
+  udDouble3 northDirection = udDouble3::create(0, 1, 0); //TODO: Fix this
+  udDouble3 up = vcGIS_GetWorldLocalUp(space, localCoords);
+
+  udDouble3 currentLatLong = udGeoZone_CartesianToLatLong(space.zone, localCoords);
+  currentLatLong.x = udClamp(currentLatLong.x, -90.0, 89.9);
+  northDirection = udNormalize(udGeoZone_LatLongToCartesian(space.zone, udDouble3::create(currentLatLong.x + 0.1, currentLatLong.y, currentLatLong.z)) - localCoords);
```
Comment (int-Frank) [inline src/vcGIS.cpp:108]: This is fine, but for the sake of whoever has to read this code (particularly math-heavy code), should we adopt a more verbose way of writing? For example
```
  slightlyNorthLatLong = udDouble3::create(currentLatLong.x + 0.1, currentLatLong.y, currentLatLong.z);
  slightlyNorthCartesian = udGeoZone_LatLongToCartesian(slightlyNorthLatLong);
  northVector = udNormalize(slightlyNorthCartesian - localCoords);
```

File: src/vcGIS.cpp [hunk 6]
```diff
@@ -94,3 +94,67 @@ udDouble3 vcGIS_GetWorldLocalUp(const vcGISSpace &space, udDouble3 localCoords)
 
   return udNormalize(upVector - localCoords);
 }
+
+udDouble3 vcGIS_GetWorldLocalNorth(const vcGISSpace &space, udDouble3 localCoords)
+{
+  if (!space.isProjected || space.zone.projection >= udGZPT_TransverseMercator)
+    return udDouble3::create(0, 1, 0);
+
+  udDouble3 northDirection = udDouble3::create(0, 1, 0); //TODO: Fix this
```
Comment (int-Frank) [inline src/vcGIS.cpp:103]: I'm not sure what there is to fix here? The algorithm seems complete without having to define `northDirection` until 108.

File: src/main.cpp [hunk 7]
```diff
@@ -1363,32 +1363,23 @@ void vcRenderSceneUI(vcState *pProgramState, const ImVec2 &windowPos, const ImVe
...
-
-        udDouble3 east = udCross(northDirection, up);
-        udDouble3 northFlat = udCross(up, east);
+        udDouble3 northDir = vcGIS_GetWorldLocalNorth(pProgramState->gis, pProgramState->camera.position);
 
         udDouble3 camRight = udCross(cameraDirection, up);
         udDouble3 camFlat = udCross(up, camRight);
 
-        double angle = udATan2(camFlat.x * northFlat.y - camFlat.y * northFlat.x, camFlat.x * northFlat.x + camFlat.y * northFlat.y);
+        double angle = udATan2(camFlat.x * northDir.y - camFlat.y * northDir.x, camFlat.x * northDir.x + camFlat.y * northDir.y);
```
Comment (int-Frank) [inline src/main.cpp:1374]: This is fine, but it is not immediately obvious what is happening here; we are using a combination of the perpendicular dot product `camFlat.x * northDir.y - camFlat.y * northDir.x` and the dot product `camFlat.x * northDir.x + camFlat.y * northDir.y` to find the angle between two vectors. This doesn't need to change but is another example where we can be a little more verbose to give clues to the reader as to what is happening at a deeper level.

File: src/vcCamera.cpp [hunk 8]
```diff
@@ -183,7 +185,7 @@ void vcCamera_Apply(vcState *pProgramState, vcCamera *pCamera, vcCameraSettings
     if (isnan(pCamera->position.x) || isnan(pCamera->position.y) || isnan(pCamera->position.z))
       pCamera->position = udDouble3::zero();
 
-    pCamInput->smoothRotation += pCamInput->mouseInput * 0.5;
+    pCamInput->smoothRotation += udDouble2::create(-pCamInput->mouseInput.x, pCamInput->mouseInput.y) * 0.5; // Because this messes with HEADING, has to be reversed
```
Comment (int-Frank) [inline src/vcCamera.cpp:188]: How does this mess with HEADING?

File: src/vcGIS.cpp [hunk 9]
```diff
@@ -94,3 +94,67 @@ udDouble3 vcGIS_GetWorldLocalUp(const vcGISSpace &space, udDouble3 localCoords)
...
+  udDouble3 up = vcGIS_GetWorldLocalUp(space, localCoords);
```
Comment (int-Frank) [inline src/vcGIS.cpp:104]: This does not actually create a normalised vector for localCoords = {0, 0, any z}. I'm adding this as a separate bug AB#1404

File: src/vcCamera.cpp [hunk 10]
```diff
@@ -155,20 +157,20 @@ void vcCamera_Apply(vcState *pProgramState, vcCamera *pCamera, vcCameraSettings
...
     // Translation
-    if (pCamSettings->lockAltitude)
-    {
-      addPos = (udDouble4x4::rotationYPR(udDouble3::create(pCamera->eulerRotation.x, 0.0, 0.0)) * udDouble4::create(addPos, 1)).toVector3();
-      addPos.z = 0.0; // might be unnecessary now
-      if (addPos.x != 0.0 || addPos.y != 0.0)
-        addPos = udNormalize3(addPos);
-    }
-    else // Lock Altitude
+    //if (pCamSettings->lockAltitude)
```
Comment (int-Frank) [inline src/vcCamera.cpp:160]: Maybe just remove this?

File: src/vcCamera.cpp [hunk 11]
```diff
@@ -243,22 +241,23 @@ void vcCamera_Apply(vcState *pProgramState, vcCamera *pCamera, vcCameraSettings
     double travelProgress = udEase(pCamInput->progress, udET_CubicInOut);
     pCamera->position = pCamInput->startPosition + moveVector * travelProgress;
 
-    udDouble3 targetEuler = udDirectionToYPR(pProgramState->worldAnchorPoint - (pCamInput->startPosition + moveVector * closest));
-    pCamera->eulerRotation = udSlerp(pCamInput->startAngle, udDoubleQuat::create(targetEuler), travelProgress).eulerAngles();
+    udDoubleQuat targetAngle = udDoubleQuat::create(pProgramState->worldAnchorPoint - (pCamInput->startPosition + moveVector * closest), 0);
```
Comment (int-Frank) [inline src/vcCamera.cpp:244]: The axis should be normalised before passing to `udDoubleQuat::create()`

**Pull request merged**